lib/otdoa_al/otdoa_nordic_at.c

```diff
@@ -0,0 +1,404 @@
+			i_ret = OTDOA_EVENT_FAIL_BAD_MODEM_RESP;
+		} else {
+			params->ecgi = u32_egci;
+			params->dlearfcn = u32_dlearfcn;
+			params->act = u32_AcT;
+			params->mcc = u16_mcc;
+			params->mnc = u16_mnc;
+			params->pci = u16_pci;
+			params->reg_status = u32_reg_status;
+		}
+	} else if (OTDOA_EVENT_FAIL_NO_DLEARFCN == i_ret && u32_egci != 0) {
+		/* return the ECGI and default DLEARFCN, and error code indicating no DLEARFCN */
+		params->ecgi = u32_egci;
+		params->dlearfcn = UNKNOWN_UBSA_DLEARFCN;
+		params->act = u32_AcT;
+		params->mcc = u16_mcc;
+		params->mnc = u16_mnc;
+		params->pci = u16_pci;
+		params->reg_status = u32_reg_status;
+	} else if (OTDOA_EVENT_FAIL_NO_PCI == i_ret && u32_egci != 0) {
+		/* return the ECGI and default DLEARFCN, and error code indicating no DLEARFCN */
+		params->ecgi = u32_egci;
+		params->dlearfcn = u32_dlearfcn;
+		params->act = u32_AcT;
+		params->mcc = u16_mcc;
+		params->mnc = u16_mnc;
+		params->pci = UNKNOWN_UBSA_PCI;
+		params->reg_status = u32_reg_status;
+	}
+
+	if (i_ret != 0
+		&& i_ret != OTDOA_EVENT_FAIL_NO_DLEARFCN && i_ret != OTDOA_EVENT_FAIL_NO_PCI) {
+		if (psz_resp) {
+			LOG_ERR("AT%%XMONITOR response %s", psz_resp);
+		}
+		LOG_ERR("Failed to parse AT%%XMONITOR response.  returning %d\n", i_ret);
+	} else {
+		LOG_DBG("otdoa_nordic_at_parse_xmonitor_response: ECGI=%" PRIu32
+			      " (0x%08X), DLEARFCN=%" PRIu32 " returning %d\n",
+			      u32_egci, (unsigned int)u32_egci, u32_dlearfcn, i_ret);
+	}
+
+	return i_ret;
+}
+
+/* Use AT%%XMONITOR command to get the current ECGI and DLEARFCN from the modem */
+int otdoa_nordic_at_get_xmonitor(otdoa_xmonitor_params_t *params)
+{
+	int i_ret = 0;
+	static char monitor_buf[256] = {0};
+
+	memset(monitor_buf, 0, sizeof(monitor_buf));
+	i_ret = nrf_modem_at_cmd(monitor_buf, sizeof(monitor_buf), "AT%%XMONITOR");
+	if (i_ret) {
+		LOG_ERR("otdoa_nordic_at_get_xmonitor: ERROR (%d) Failed to get "
+			      "MODEM Status\n",
+			      i_ret);
+		i_ret = OTDOA_EVENT_FAIL_BAD_MODEM_RESP;
+	} else {
+		i_ret = otdoa_nordic_at_parse_xmonitor_response(monitor_buf,
+						       strlen(monitor_buf), params);
+	}
+	return i_ret;
+}
+
+/*
+ * The AT command query we will use.  Note that "=1" ensures we get the IMEI
+ * value (see AT command reference)
+ */
+#define IMEI_QUERY	     "AT+CGSN=1"
+#define IMEI_QUERY_LEN	     sizeof(IMEI_QUERY)
+#define IMEI_RESP	     "+CGSN: \""
+#define IMEI_RESP_LEN	     sizeof(IMEI_RESP)
+#define NRF_IMEI_LEN	     15
+#define CGSN_RESPONSE_LENGTH 35
+static char otdoa_nordic_at_imei[NRF_IMEI_LEN + 1] = {0}; /* "+1" to ensure null term */
+
+/**
+ * Sends AT+CGSN=1
+ * Resp:
+ * +CGSN: "352656100367872"
+ * OK
+ *
+ * Where the IMEI is a 15-digit string
+ */
+int otdoa_nordic_at_get_imei_from_modem(void)
+{
+	char imei_buf[CGSN_RESPONSE_LENGTH + 1];
+	int err;
+
+	memset(otdoa_nordic_at_imei, 0, sizeof(otdoa_nordic_at_imei));
+
+	err = nrf_modem_at_cmd(imei_buf, sizeof(imei_buf), IMEI_QUERY);
```

Comment:
There is a `nrf_modem_at_scanf()` function that can be used to `scanf` a response dicrectly.
I think here you could have used something like:
```
nrf_modem_at_scanf(IMEI_QUERY, "+CGSN: \"%35[0-9]\"", imei_buf);
```
It's perfectly fine to parse manually, especially for large responses, but for simple stuff this may just be enough. Just FYI.